src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp

```diff
@@ -0,0 +1,186 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+
+#include <array>
+#include <cstddef>
+#include <functional>
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Transpose.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+void apply_matrix_in_first_dim(double* result, const double* const input,
+                               const Matrix& matrix,
+                               const size_t size) noexcept {
+  dgemm_<true>('N', 'N',
+               matrix.rows(),            // rows of matrix and result
+               size / matrix.columns(),  // columns of result and input
+               matrix.columns(),         // columns of matrix and rows of input
+               1.0,                      // overall multiplier
+               matrix.data(),            // matrix
+               matrix.spacing(),         // rows of matrix including padding
+               input,                    // input
+               matrix.columns(),         // rows of input
+               0.0,                      // overwrite output with result
+               result,                   // result
+               matrix.rows());           // rows of result
+}
+}  // namespace
+
+template <typename SymmList, typename IndexList, size_t Dim>
+void logical_partial_derivative(
+    const gsl::not_null<TensorMetafunctions::prepend_spatial_index<
+        Tensor<DataVector, SymmList, IndexList>, Dim, UpLo::Lo,
+        Frame::Logical>*>
+        logical_derivative_of_u,
+    const gsl::not_null<gsl::span<double>*> buffer,
+    const Tensor<DataVector, SymmList, IndexList>& u,
+    const Mesh<Dim>& mesh) noexcept {
+  static_assert(
+      Dim > 0 and Dim < 4,
+      "logical_partial_derivative is only implemented for 1, 2, and 3d");
+  const size_t num_grid_points = mesh.number_of_grid_points();
+  ASSERT(buffer->size() >= num_grid_points,
+         "The buffer in logical_partial_derivative must be at least of size "
+             << num_grid_points << " but is of size " << buffer->size());
+
+  destructive_resize_components(logical_derivative_of_u,
+                                mesh.number_of_grid_points());
+  const Matrix identity{};
```

Comment:
Is this somehow actually set to an identity matrix of some dimension? If not, perhaps a name like `empty_matrix` would be better?

Reply:
Agreed. For `apply_matrices` an empty matrix is treated as the identity matrix, but here it doesn't fulfill that role.